### PR TITLE
Fix building on LLVM libc++

### PR DESCRIPTION
**Pull request by JamiKettunen**
This should complete the first part of my journey of running Mir on [Chimera Linux](https://chimera-linux.org) initially reported on #3383.

Now I'm figuring out this runtime failure of `mir_demo_server` (https://paste.c-net.org/umtbetmnnvsn) which I cannot reproduce on Arch with (https://paste.c-net.org/a3azgse7x8en) or without these patches, further opinions welcome :)

### CODE REVIEW

File: src/server/console/minimal_console_services.cpp [hunk 1]
```diff
@@ -100,7 +100,7 @@ std::future<std::unique_ptr<mir::Device>> mir::MinimalConsoleServices::acquire_d
     auto const devnode = [](auto const& fd) {
         using namespace boost::iostreams;
         char line_buffer[1024];
-        stream<file_descriptor_source> uevent{fd, file_descriptor_flags::never_close_handle};
+        stream<file_descriptor_source> uevent{(int)fd, file_descriptor_flags::never_close_handle};
```
Comment (AlanGriffiths) [inline src/server/console/minimal_console_services.cpp:103]: `Fd` has an implicit conversion to `int`, so I'm surprised this is needed. But it should not be a c-style cast.

File: include/miral/miral/output.h [hunk 2]
```diff
@@ -108,7 +108,7 @@ class Output
 
     /// A custom attribute map
     /// \remark Since MirAL 3.8
-    auto attributes_map() const -> std::map<std::string const, std::optional<std::string>>;
+    auto attributes_map() const -> std::map<std::string, std::optional<std::string>>;
```
Comment (AlanGriffiths) [inline include/miral/miral/output.h:111]: Removing the old signature is an ABI break. We should retain that.

File: src/miral/static_display_config.h [hunk 3]
```diff
@@ -69,7 +69,7 @@ class YamlFileDisplayConfig : public mir::graphics::DisplayConfigurationPolicy
         mir::optional_value<float>  scale;
         mir::optional_value<MirOrientation>  orientation;
         mir::optional_value<int> group_id;
-        std::map<std::string const, std::optional<std::string>> custom_attribute;
+        std::map<std::string, std::optional<std::string>> custom_attribute;
```
Comment (AlanGriffiths) [inline src/miral/static_display_config.h:72]: Changing this type is fine

File: src/server/console/linux_virtual_terminal.cpp [hunk 4]
```diff
@@ -655,7 +655,7 @@ std::future<std::unique_ptr<mir::Device>> mir::LinuxVirtualTerminal::acquire_dev
         {
             using namespace boost::iostreams;
             char line_buffer[1024];
-            stream<file_descriptor_source> uevent{fd, file_descriptor_flags::never_close_handle};
+            stream<file_descriptor_source> uevent{(int)fd, file_descriptor_flags::never_close_handle};
```
Comment (AlanGriffiths) [inline src/server/console/linux_virtual_terminal.cpp:658]: `Fd` has an implicit conversion to `int`, so I'm surprised this is needed. But it should not be a c-style cast.

File: src/platforms/virtual/display.h [hunk 5]
```diff
@@ -17,6 +17,7 @@
 #ifndef MIR_GRAPHICS_VIRT_DISPLAY_H_
 #define MIR_GRAPHICS_VIRT_DISPLAY_H_
 
+#include <mutex>
```
Comment (AlanGriffiths) [inline src/platforms/virtual/display.h:20]: We keep standard headers after project ones

File: src/server/compositor/multi_monitor_arbiter.cpp [hunk 6]
```diff
@@ -17,6 +17,7 @@
 #include "multi_monitor_arbiter.h"
 #include "mir/graphics/buffer.h"
 #include "mir/frontend/event_sink.h"
+#include <algorithm>
```
Comment (AlanGriffiths) [inline src/server/compositor/multi_monitor_arbiter.cpp:20]: We keep standard headers after project ones

**Review by AlanGriffiths — CHANGES_REQUESTED** [review]
Thanks for this, there's a little nitpicking but it is basically fine

File: include/miral/miral/output.h (hunk 2 above)
Comment (JamiKettunen) [reply include/miral/miral/output.h:111]: How do you propose that is done? If I simply add the new one right below it FTBFS with:
```
In file included from ../src/miral/active_outputs.cpp:18:
In file included from ../src/miral/display_configuration_listeners.h:20:
../include/miral/miral/output.h:112:10: error: functions that differ only in their return type cannot be overloaded
  112 |     auto attributes_map() const -> std::map<std::string, std::optional<std::string>>;
      |          ^
../include/miral/miral/output.h:111:10: note: previous declaration is here
  111 |     auto attributes_map() const -> std::map<std::string const, std::optional<std::string>>;
      |          ^
```

Comment (AlanGriffiths) [reply include/miral/miral/output.h:111]: What problem did you see that motivated you to change the return type?

Comment (JamiKettunen) [reply include/miral/miral/output.h:111]: #3383 basically, if I don't do each hunk of the changes in 517b4b13a1bc033fd4fd03f453157d8f88ef7389 I see a similar error as I originally reported there

Comment (JamiKettunen) [reply include/miral/miral/output.h:111]: To be specific, without the `attributes_map` (`include/miral/miral/output.h` and `src/miral/output.cpp`) changes I get:
```
../src/miral/output.cpp:127:12: error: no viable conversion from returned value of type 'map<basic_string<...>, [...]>' to function return type 'map<const basic_string<...>, [...]>'
  127 |     return self->custom_attribute;
      |            ^~~~~~~~~~~~~~~~~~~~~~
/usr/bin/../include/c++/v1/map:1078:25: note: candidate constructor not viable: no known conversion from 'std::map<std::string, std::optional<std::string>>' (aka 'map<basic_string<char>, optional<basic_string<char>>>') to 'const map<const string, optional<string>> &' for 1st argument
 1078 |   _LIBCPP_HIDE_FROM_ABI map(const map& __m) : __tree_(__m.__tree_) { insert(__m.begin(), __m.end()); }
      |                         ^   ~~~~~~~~~~~~~~
/usr/bin/../include/c++/v1/map:1096:25: note: candidate constructor not viable: no known conversion from 'std::map<std::string, std::optional<std::string>>' (aka 'map<basic_string<char>, optional<basic_string<char>>>') to 'map<const string, optional<string>> &&' for 1st argument
 1096 |   _LIBCPP_HIDE_FROM_ABI map(map&& __m) _NOEXCEPT_(is_nothrow_move_constructible<__base>::value)
      |                         ^   ~~~~~~~~~
/usr/bin/../include/c++/v1/map:1106:25: note: candidate constructor not viable: no known conversion from 'std::map<std::string, std::optional<std::string>>' (aka 'map<basic_string<char>, optional<basic_string<char>>>') to 'initializer_list<value_type>' (aka 'initializer_list<pair<const std::string, std::optional<std::string>>>') for 1st argument
 1106 |   _LIBCPP_HIDE_FROM_ABI map(initializer_list<value_type> __il, const key_compare& __comp = key_compare())
      |                         ^   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/bin/../include/c++/v1/map:1034:34: note: explicit constructor is not a candidate
 1034 |   _LIBCPP_HIDE_FROM_ABI explicit map(const key_compare& __comp) _NOEXCEPT_(
      |                                  ^
/usr/bin/../include/c++/v1/map:1128:34: note: explicit constructor is not a candidate
 1128 |   _LIBCPP_HIDE_FROM_ABI explicit map(const allocator_type& __a) : __tree_(typename __base::allocator_type(__a)) {}
      |                                  ^

```

Comment (AlanGriffiths) [reply include/miral/miral/output.h:111]: It should be enough to change:
```c++
  return self->custom_attribute;
```
to 
```c++
   return {self->custom_attribute.begin(), self->custom_attribute.end()};
```

File: src/server/console/minimal_console_services.cpp (hunk 1 above)
Comment (JamiKettunen) [reply src/server/console/minimal_console_services.cpp]: Switched to `static_cast<int>(fd)`

File: include/miral/miral/output.h (hunk 2 above)
Comment (JamiKettunen) [reply include/miral/miral/output.h]: Indeed that appears to do the trick

**Review by AlanGriffiths — APPROVED** [review]
Looks good now, thanks!

**Pull request merged**